File: ucf_pt/local2024QualifyingRoundData (2)/solutions/income.py
# Arup Guha
# 8/2/2024
# Solution to 2024 UCF Locals Qualifying Round Problem: Income Inequality

def solve(money):

    # Sort from big to small.
    n = len(money)
    money.sort(reverse = True)
    tot = sum(money)

    # Our initial answer, running sum.
    res = 0
    curS = 0

    # Loop through rev. sorted incomes.
    for i in range(n):

        # Sum of top i+1 people.
        curS += money[i]

        # This is what we're trying to maximize.
        res = max(res, curS/tot - (i+1)/n)

    # As a percentage.
    return 100*res

# Run it.
n = int(input())
vals = [int(x) for x in input().split()]
print(solve(vals))
    

File: ucf_pt/local2024QualifyingRoundData (2)/solutions/income/income.java
// Arup Guha
// 8/12/2024
// Alternate Solution to UCF 2024 Qualification Problem: Income Inequality

import java.util.*;

public class income {

	public static void main(String[] args) {
	
		Scanner stdin = new Scanner(System.in);
		int n = stdin.nextInt();
		
		// Read numbers.
		ArrayList<Long> vals = new ArrayList<Long>();
		long tot = 0;
		for (int i=0; i<n; i++) {
			long tmp = stdin.nextLong();
			tot += tmp;
			vals.add(tmp);
		}
		
		// Sort in reverse order.
		Collections.sort(vals);
		Collections.reverse(vals);
		double res = 0;
		long curT = 0;
		
		// Try each split of the i+1 richest people.
		for (int i=0; i<n; i++) {
			curT += vals.get(i);
			double tmp = (double)curT/tot - 1.0*(i+1)/n;
			res = Math.max(res, 100*tmp);
		}
		
		// Ta da!
		System.out.println(res);
	}
}